sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUuidPathColumnOnProjects.java

```diff
@@ -0,0 +1,156 @@
+    massUpdate.update("update projects set uuid_path=? where uuid=? and uuid_path is null");
+    massUpdate.rowPluralName("components");
+    massUpdate.execute((row, update) -> handleComponent(relations, row, update));
+  }
+
+  private void handleOrphans(Context context) throws SQLException {
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("select uuid, project_uuid from projects where uuid_path is null");
+    massUpdate.update("update projects set uuid_path=? where uuid=? and uuid_path is null");
+    massUpdate.rowPluralName("orphans");
+    massUpdate.execute((row, update, updateIndex) -> {
+      String uuid = row.getString(1);
+      String rootUuid = row.getString(2);
+      String path = uuid.equals(rootUuid) ? ROOT_PATH : (rootUuid + PATH_SEPARATOR);
+      update.setString(1, path);
+      update.setString(2, uuid);
+      return true;
+    });
+  }
+
+  private boolean handleComponent(Relations relations, Select.Row row, SqlStatement update) throws SQLException {
+    String componentUuid = row.getString(1);
+    String rootComponentUuid = row.getString(2);
+
+    if (componentUuid.equals(rootComponentUuid)) {
+      // Root component, no need to use the table SNAPSHOTS.
+      // Moreover it allows to support provisioned projects (zero analysis)
+      update.setString(1, PATH_SEPARATOR);
+      update.setString(2, componentUuid);
+      return true;
+    }
+
+    Snapshot snapshot = relations.snapshotsByComponentUuid.get(componentUuid);
+    if (snapshot == null) {
+      LOG.trace("No UUID found for component UUID={}", componentUuid);
+      return false;
+    }
+
+    List<String> componentUuidPath = snapshot.snapshotPath.stream()
+      .map(snapshotId -> relations.snapshotsById.get(snapshotId).componentUuid)
+      .collect(toCollection(() -> new ArrayList<>(snapshot.snapshotPath.size())));
+    update.setString(1, PATH_JOINER.join(componentUuidPath) + PATH_SEPARATOR);
+    update.setString(2, componentUuid);
+    return true;
+  }
+
+  private static final class Relations {
+    private final Map<String, Snapshot> snapshotsByComponentUuid = new HashMap<>();
+    private final Map<Long, Snapshot> snapshotsById = new HashMap<>();
+
+    void add(Snapshot snapshot) {
+      snapshotsByComponentUuid.put(snapshot.componentUuid, snapshot);
+      snapshotsById.put(snapshot.id, snapshot);
+    }
+  }
+
+  private static final class Snapshot {
+    private final long id;
+    private final List<Long> snapshotPath;
+    private final String componentUuid;
+
+    public Snapshot(long id, String snapshotPath, String componentUuid) {
+      this.id = id;
+      this.snapshotPath = parsePath(snapshotPath);
+      this.componentUuid = componentUuid;
+    }
+
+    // inputs: "", "1." or "1.2.3."
+    private List<Long> parsePath(String snapshotPath) {
+      return PATH_SPLITTER
+        .splitToList(snapshotPath)
+        .stream()
+        .map(Long::parseLong)
```

Comment:
did you know you can use `mapToLong` instead of `map`, store in a `long[]` and save (some) memory ?
Definitely not necessary here but just saying :)